src/gallium/drivers/lima/lima_draw.c

```diff
@@ -908,6 +916,31 @@ lima_draw_vbo(struct pipe_context *pctx, const struct pipe_draw_info *info)
       ctx->const_buffer[PIPE_SHADER_FRAGMENT].dirty = false;
    }
 
+   if (!ctx->num_draws ||
+      (ctx->dirty & LIMA_CONTEXT_DIRTY_TEXTURES)) {
+      lima_update_textures(ctx);
+      if (lima_dump_command_stream) {
+         unsigned tex_desc_oft = pp_tex_descs_offset +
+                              ctx->buffer_state[lima_ctx_buff_pp_tex_desc].offset;
+         printf("lima: add textures_desc at va %x\n",
+                 ctx->pp_buffer->va + tex_desc_oft);
+         lima_dump_blob(ctx->pp_buffer->map + tex_desc_oft,
+                        ctx->buffer_state[lima_ctx_buff_pp_tex_desc].size,
+                        false);
+      }
+   }
+
+   if (ctx->tex_stateobj.num_samplers) {
```

Comment:
Can we do submit_add_bo in lima_update_textures too?

Reply:
Do we need to submit them each draw even if they were not changed? If yes, they have to be in draw_vbo(), otherwise I can move them to lima_update_textures.

Reply:
No, just need to add a BO once every frame.

Reply:
Right, so if textures were not updated in this frame (and LIMA_CONTEXT_DIRTY_TEXTURES is not set) we still need to submit them. But lima_update_textures() won't be called in this case.

Reply:
!ctx->num_draws make sure every first draw in a frame will call lima_update_textures.